src/algebra/char_p/basic.lean

```diff
@@ -80,19 +82,35 @@ let ⟨c, H⟩ := char_p.exists α in ⟨c, H, λ y H2, char_p.eq α H2 H⟩
 noncomputable def ring_char (α : Type u) [semiring α] : ℕ :=
 classical.some (char_p.exists_unique α)
 
-theorem ring_char.spec (α : Type u) [semiring α] : ∀ x:ℕ, (x:α) = 0 ↔ ring_char α ∣ x :=
+namespace ring_char
+
+theorem spec (α : Type u) [semiring α] : ∀ x:ℕ, (x:α) = 0 ↔ ring_char α ∣ x :=
 by letI := (classical.some_spec (char_p.exists_unique α)).1;
 unfold ring_char; exact char_p.cast_eq_zero_iff α (ring_char α)
 
-theorem ring_char.eq (α : Type u) [semiring α] {p : ℕ} (C : char_p α p) : p = ring_char α :=
+theorem eq (α : Type u) [semiring α] {p : ℕ} (C : char_p α p) : p = ring_char α :=
 (classical.some_spec (char_p.exists_unique α)).2 p C
 
-theorem add_pow_char_of_commute (R : Type u) [ring R] {p : ℕ} [fact p.prime]
+instance char_p (R : Type u) [semiring R] : char_p R (ring_char R) :=
+⟨spec R⟩
+
+theorem of_eq {R : Type u} [semiring R] {p : ℕ} (h : ring_char R = p) : char_p R p :=
```

Comment:
Would it make sense to generalise `ring_char R` to `q` (given the instance above this statement)?

Reply:
I don't understand. What is the alternative statement?

Reply:
Maybe it should be called `char_p.congr`:
```
theorem congr {R : Type u} [semiring R] {p : ℕ} (q : ℕ) [char_p R q] (h : q = p) : char_p R p :=
```

Reply:
But then why can't I just rewrite?

Reply:
Why can't you rewrite instead of the current statement?

Reply:
Done